From 66294cf3338ae124c762dbfdb944ed5379627ace Mon Sep 17 00:00:00 2001
From: Cheng Jian <cj.chengjian@huawei.com>
Date: Mon, 29 Nov 2021 16:28:53 +0800
Subject: [PATCH] lib/clear_user: ensure loop in __arch_clear_user
 cache-aligned v2
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 66294cf3338ae124c762dbfdb944ed5379627ace
Modified-by-SEL: Yes, refreshed due to different context


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I4K6CJ
CVE: NA

--------------------------------

We must ensure that the following four instructions are cache-aligned.
Otherwise, it will cause problems with the performance of libMicro
pread.

1:
        # uao_user_alternative 9f, str, sttr, xzr, x0, 8
        str     xzr, [x0], #8
        nop
        subs    x1, x1, #8
        b.pl    1b

with this patch:

             prc thr   usecs/call      samples   errors cnt/samp     size
pread_z100     1   1      5.88400          807        0 1            102400

The result of pread can range from 5 to 9 depending on  the
alignment performance of this function.

Signed-off-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/lib/clear_user.S | 3 +++
 1 file changed, 3 insertions(+)

--- a/arch/arm64/lib/clear_user.S
+++ b/arch/arm64/lib/clear_user.S
@@ -23,6 +23,9 @@ SYM_FUNC_START(__arch_clear_user)
 	mov	x2, x1			// save the size for fixup return
 	subs	x1, x1, #8
 	b.mi	2f
+#ifdef CONFIG_ARCH_HISI
+	.align 5
+#endif
 1:
 user_ldst 9f, sttr, xzr, x0, 8
 	subs	x1, x1, #8
